.changeset/purple-pans-shout.md

```diff
@@ -0,0 +1,537 @@
+---
+"@atproto/lex-cli": minor
+"@atproto/xrpc": minor
+"@atproto/api": minor
+"@atproto/dev-env": patch
+"@atproto/lexicon": patch
+"@atproto/ozone": patch
+"@atproto/bsky": patch
+"@atproto/pds": patch
+"@atproto/xrpc-server": patch
+---
+
+## New Features
+
+1) We've restructured the `XrpcClient` HTTP fetch handler to be specified during
+   the instantiation of the XRPC client, through the constructor, instead of
+   using a default - and statically defined - implementation.
+2) With this refactor, the XRPC client is now more adaptable to various use
+   cases. In particular, session management, retries, and other request-specific
+   logic (signing, etc.) can be implemented in the fetch handler itself rather
+   than by the calling code.
+3) A new abstract class named `Agent`, has been added to `@atproto/api`. This
+   class is the base class for all Bluesky agents classes in the `@atproto`
+   ecosystem. It is meant to be extended by implementation that provide
+   session management and fetch handling for the `AtpClient` instances.
+
+## Motivation
+
+This change is motivated by the need to make the `@atproto/api` package more
+flexible when it comes to session management. In particular, they are forward
+compatible with the upcoming support of OAuth by atproto.
+
+In addition to this, the redesigned session management system fixes a bug that
+could cause the session data to become invalid when Agent clones are created
+(e.g. using `agent.withProxy()`).
+
+## Non-breaking changes
+
+- The `com.*` and `app.*` namespaces have been made directly available
+  to every `Agent` instances.
+
+## Breaking changes
+
+Most of the changes introduced in this version are backward-compatible. However,
+there are a couple of breaking changes you should be aware of:
+
+- Customizing `fetch`: The ability to customize the `fetch: FetchHandler`
+  property of `@atproto/xrpc`'s `Client` and `@atproto/api`'s `AtpAgent` classes
+  has been removed. Previously, the `fetch` property could be set to a function
+  that would be used as the fetch handler for that instance, and was initialized
+  to a default fetch handler. That property is still accessible in a read-only
+  fashion through the `fetchHandler` property. That property can only be set
+  during the instance creation. Attempting to set/get the `fetch` property will
+  now result in an error.
+
+- The `fetch()` method, as well as WhatWG compliant `Request` and `Headers`
+  constructors, must be globally available in your environment. Use a polyfill
+  if necessary.
+
+- The `AtpBaseClient` has been removed. The `AtpServiceClient` has been renamed
+  `AtpBaseClient`. Any code using either of these classes will need to be
+  updated.
```

Comment:
This reads a little confusing: there is still a `AtpBaseClient`, but the old implementation was removed and replaced with what was formerly called `AtpServiceClient`?

Reply:
Yes! I guess I could revert the renaming ...

Reply:
I don't mean change the code, just re-phrase this sentence.